// webpack-preprocessor-loader/dist/preprocessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.preprocessor = void 0;
const filter_1 = require("./filter");
const global_1 = require("./global");
const printer_1 = require("./printer");
const reader_1 = require("./reader");
/**
 * The preprocessor
 *
 * @export
 * @param {IWebpackLoaderContext} this webpack loader context
 * @param {string} content raw text file
 * @returns {string}
 */
function preprocessor(content) {
    const { directives, params, verbose } = getOptions(this.query);
    const r = reader_1.reader(content);
    const f = filter_1.filter(directives, params);
    const p = printer_1.printer(verbose);
    f.next();
    p.next();
    const o = {
        block: '',
        raw: '',
        c_open: '',
        c_close: '',
        is_comment: false,
        is_directive: false,
        is_keep: true,
    };
    for (const reader_state of r) {
        const filter_state = f.next(reader_state).value;
        p.next(Object.assign(o, reader_state, filter_state));
    }
    return p.next({}).value || '';
}
exports.preprocessor = preprocessor;
function getOptions(query) {
    const options = Object.assign({}, global_1.DEFAULT_OPTIONS, query);
    options.directives.debug = options.debug;
    return options;
}
